File: libtungsten/usbcom/usbcom.py
import usb.core, usb.util
import threading

class USBCom:
    OUT = 0
    IN = 1

    REQ_BOOTLOADER = 0x00
    REQ_CONNECT = 0x01
    REQ_DISCONNECT = 0x02

    def __init__(self):
        self.connected = False
        self.connection_error = False
        self.connection_error_message = ""

        self._dev = None
        self._ep_in = None
        self._ep_out = None

        self._read_handler = None
        self._thread_usb = None

    def connect(self, read_handler):
        # Reset state
        if self.connected and not self.connection_error:
            self.disconnect()
        self.connected = False
        self.connection_error = False
        self.connection_error_message = ""

        # Look for an USB device with matching ids
        self._dev = usb.core.find(idVendor=0x03eb, idProduct=0xcabd);
        if self._dev is None:
            self.connection_error = True
            self.connection_error_message = "Device not found"
            for listener in self.listeners:
                listener._on_connection_status(False, self.connection_error_message)
            return False

        # Configure the device
        self._dev.set_configuration()
        cfg = self._dev.get_active_configuration()
        intf = cfg[(0, 0)]

        # Find the IN and OUT endpoints
        self._ep_in = usb.util.find_descriptor(intf, custom_match=lambda e: usb.util.endpoint_direction(e.bEndpointAddress) == usb.util.ENDPOINT_IN)
        self._ep_in = usb.util.find_descriptor(intf, custom_match=lambda e: usb.util.endpoint_direction(e.bEndpointAddress) == usb.util.ENDPOINT_IN)
        self._ep_out = usb.util.find_descriptor(intf, custom_match=lambda e: usb.util.endpoint_direction(e.bEndpointAddress) == usb.util.ENDPOINT_OUT)
        assert self._ep_in is not None
        assert self._ep_out is not None

        # Connected
        self.connected = True
        self._send_request_internal(USBCom.REQ_CONNECT)

        # Start the read thread
        self._read_handler = read_handler
        self._thread_usb = threading.Thread(target=self._usb_handler)
        self._thread_usb.daemon = True
        self._thread_usb.start()

        return True

    def disconnect(self):
        if self.connected and not self.connection_error:
            self._send_request_internal(USBCom.REQ_DISCONNECT)
            self.connected = False
            self._thread_usb.join()

    def _send_request_internal(self, request, value=0, index=0, direction=OUT, payload=None):
        if self.connected and not self.connection_error:
            bmRequestType = direction << 7 | 2 << 5
            self._dev.ctrl_transfer(bmRequestType, request, value, index, payload)

    def send_request(self, request, value=0, index=0, direction=OUT, payload=None):
        return self._send_request_internal(request | 0x80, value, index, direction, payload)

    def write(self, data):
        if self.connected and not self.connection_error:
            self._ep_out.write(data)

    def _usb_handler(self):
        while self.connected and not self.connection_error:
            try:
                buff = self._ep_in.read(64, 100)
            except usb.core.USBError as e:
                if e.errno != 110:
                    self.connected = False
                    self.connection_error = True
                    self.connection_error_message = str(e)
                    print("")
                    print("USB error " + str(e.errno) + " : " + e.strerror)
                    return
            else:
                if len(buff) > 0:
                    self._read_handler(buff)

File: libtungsten/usbcom/USBCom.cpp
#include "USBCom.h"

#include <iostream>

const std::string LIBUSB_ERROR_STRINGS[] = {
    "SUCCESS",
    "ERROR_IO",
    "ERROR_INVALID_PARAM",
    "ERROR_ACCESS",
    "ERROR_NO_DEVICE",
    "ERROR_NOT_FOUND",
    "ERROR_BUSY",
    "ERROR_TIMEOUT",
    "ERROR_OVERFLOW",
    "ERROR_PIPE",
    "ERROR_INTERRUPTED",
    "ERROR_NO_MEM",
    "ERROR_NOT_SUPPORTED"
};

const uint8_t REQ_BOOTLOADER = 0x00;
const uint8_t REQ_CONNECT = 0x01;
const uint8_t REQ_DISCONNECT = 0x02;


int USBCom::init() {
    int r = 0;

    // Init libusb
    r = libusb_init(NULL);
    if (r < 0) {
        printLibUSBError("Unable to initialize libusb", r);
        return r;
    }

    // Find a matching usb device
    r = findDevice(DEFAULT_USB_VENDOR_ID, DEFAULT_USB_PRODUCT_ID);
    if (r < 0) {
        return r;
    }

    // Set the device's default configuration
    r = libusb_set_configuration(_handle, 0);
    if (r < 0) {
        printLibUSBError("libusb_set_configuration failed", r);
        libusb_close(_handle);
        return r;
    }

    // Remove any active driver to take control of the interface
    if (libusb_kernel_driver_active(_handle, DEFAULT_INTERFACE)) {
        r = libusb_detach_kernel_driver(_handle, DEFAULT_INTERFACE);
        if (r < 0) {
            printLibUSBError("Unable to detach kernel driver", r);
            libusb_close(_handle);
            return r;
        }
    }

    // Claim the interface
    r = libusb_claim_interface(_handle, DEFAULT_INTERFACE);
    if (r < 0) {
        printLibUSBError("Unable to claim device's interface", r);
        libusb_close(_handle);
        return r;
    }

    sendRequestInternal(REQ_CONNECT);

    return LIBUSB_SUCCESS;
}

void USBCom::close() {
    if (_handle == nullptr) {
        return;
    }
    sendRequestInternal(REQ_DISCONNECT);
    libusb_close(_handle);
    _handle = nullptr;
}

int USBCom::read(uint8_t* buffer, int maxLength, int timeout) {
    int transferredLength = 0;
    int r = libusb_bulk_transfer(_handle, LIBUSB_ENDPOINT_IN | 1, buffer, maxLength, &transferredLength, timeout);
    if (r < 0) {
        printLibUSBError("Error during read", r);
        exit(r);
    }
    return transferredLength;
}

int USBCom::write(const uint8_t* buffer, int length, int timeout) {
    int transferredLength = 0;
    int r = libusb_bulk_transfer(_handle, 2, (uint8_t*)buffer, length, &transferredLength, timeout);
    if (r < 0) {
        printLibUSBError("Error during read", r);
        exit(r);
    }
    return transferredLength;
}

int USBCom::sendRequest(uint8_t request, uint16_t value, uint16_t index, Direction direction, uint8_t* buffer, uint16_t length, int timeout) {
    return USBCom::sendRequestInternal(request | 0x80, value, index, direction, buffer, length, timeout);
}

int USBCom::sendRequestInternal(uint8_t request, uint16_t value, uint16_t index, Direction direction, uint8_t* buffer, uint16_t length, int timeout) {
    uint8_t bmRequestType
        = static_cast<int>(direction) << 7  // Direction
        | 2 << 5                            // Type : vendor
        | 0;                                // Recipent : device
    int r = libusb_control_transfer(_handle, bmRequestType, request, value, index, buffer, length, timeout);
    if (r < 0) {
        printLibUSBError("Error during request transfer", r);
        exit(0);
    }
    return r;
}




int USBCom::findDevice(uint16_t vid, uint16_t pid) {
    int r = 0;
    int returnCode = 0;

    // Get the list of devices
    libusb_device** list;
    libusb_device* board = nullptr;
    ssize_t cnt = libusb_get_device_list(NULL, &list);
    if (cnt < 0) {
        printLibUSBError("Unable to find the list of usb devices", r);
        return r;
    }

    // Look for a device with the matching vendor and product ids
    for (ssize_t i = 0; i < cnt; i++) {
        libusb_device *device = list[i];
        struct libusb_device_descriptor desc;
        r = libusb_get_device_descriptor(device, &desc);
        if (r < 0) {
            std::cerr << "Warning : unable to get descriptor for device (bus "
                      << libusb_get_bus_number(device) << ", device " << libusb_get_device_address(device) << ")" << std::endl;
            continue;
        }
        if (desc.idVendor == vid && desc.idProduct == pid) {
            // Found matching device
            if (board == nullptr) {
                board = device;
            } else {
                std::cout << "Warning : more than one matching device found, are there multiple boards plugged in? The first match will be used." << std::endl;
                break;
            }
        }
    }

    if (board != nullptr) {
        // Open the device
        r = libusb_open(board, &_handle);
        if (r < 0) {
            //std::cerr << "Unable to open device 0x" << std::hex << vid << ":0x" << pid << std::dec << " : error " << r << std::endl;
            printLibUSBError("Unable to open device", r);
            if (r == LIBUSB_ERROR_IO) {
                std::cerr << "Please try again" << std::endl;
            }
            returnCode = r;
        }
    } else {
        returnCode = LIBUSB_ERROR_NO_DEVICE;
    }

    // Free the device list and return true if a device was found and opened correctly
    libusb_free_device_list(list, 1);
    return returnCode;
}

void USBCom::printLibUSBError(std::string message, int r) {
    std::cerr << message << " : ";
    if (-r <= 12) {
        std::cerr << LIBUSB_ERROR_STRINGS[-r] << std::endl;
    } else {
        std::cerr << "error " << r << std::endl;
    }
}